app/src/main/java/org/schabi/newpipe/player/Player.java

```diff
@@ -4234,4 +4242,86 @@ public PlayQueueAdapter getPlayQueueAdapter() {
     }
 
     //endregion
+
+
+
+    //region SurfaceHandlerCallback
+    /**
+     * Prevent error message: 'Unrecoverable player error occurred'
+     * In case of rotation some users see this kind of an error which is preventable
+     * having a Callback that handles the lifecycle of the surface.
+     *
+     * How?: In case we are no longer able to write to the surface eg. through rotation/putting in
+     * background we set set a DummySurface. Although it it works on API >= 23 only.
+     * Result: we get a little video interruption (audio is still fine) but we won't get the
+     * 'Unrecoverable player error occurred' error message.
+     *
+     * This implementation is based on:
+     *  'ExoPlayer stuck in buffering after re-adding the surface view a few time #2703'
+     *
+     *  -> exoplayer fix suggestion link
+     *   https://github.com/google/ExoPlayer/issues/2703#issuecomment-300599981
+     */
+    private final class SurfaceHolderCallback implements SurfaceHolder.Callback {
+
+        private final SimpleExoPlayer player;
+        private DummySurface dummySurface;
+
+        SurfaceHolderCallback(final SimpleExoPlayer player) {
+            this.player = player;
+        }
+
+        @Override
+        public void surfaceCreated(final SurfaceHolder holder) {
+            player.setVideoSurface(holder.getSurface());
+        }
+
+        @Override
+        public void surfaceChanged(final SurfaceHolder holder,
+                                   final int format,
+                                   final int width,
+                                   final int height) {
+        }
+
+        @Override
+        public void surfaceDestroyed(final SurfaceHolder holder) {
+            if (dummySurface == null) {
+                dummySurface = DummySurface.newInstanceV17(context, false);
+            }
+            player.setVideoSurface(dummySurface);
+        }
+
+        public void release() {
+            if (dummySurface != null) {
+                dummySurface.release();
+                dummySurface = null;
+            }
+        }
+    }
+
+    private void setupVideoSurface() {
```

Comment:
Can we be completely completely sure that this isn't called twice? That would create leaks and certainly other issues. Maybe call `cleanupVideoSurface()` just at the beginning of `setupVideoSurface()`, to ensure there is no `Surface` left behind.

Reply:
done